x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java

```diff
@@ -0,0 +1,213 @@
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.support.WriteRequest;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.client.internal.OriginSettingClient;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.ReleasableBytesStreamOutput;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.indices.ExecutorNames;
+import org.elasticsearch.indices.SystemIndexDescriptor;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.sort.SortOrder;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Collections;
+
+import static org.elasticsearch.xcontent.ToXContent.EMPTY_PARAMS;
+import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
+
+/**
+ * A service that manages the persistent {@link Engine} configurations.
+ */
+public class EngineIndexService {
+    private static final Logger logger = LogManager.getLogger(EngineIndexService.class);
+    public static final String ENGINE_ALIAS_NAME = ".engine";
+    public static final String ENGINE_CONCRETE_INDEX_NAME = ".engine-1";
+    public static final String ENGINE_INDEX_NAME_PATTERN = ".engine-*";
+    public static final String ENGINE_ORIGIN = "engine";
+
+    private final Client clientWithOrigin;
+    private final BigArrays bigArrays;
+
+    public EngineIndexService(Client client, BigArrays bigArrays) {
+        this.clientWithOrigin = new OriginSettingClient(client, ENGINE_ORIGIN);
+        this.bigArrays = bigArrays;
+    }
+
+    /**
+     * Returns the {@link SystemIndexDescriptor} for the {@link Engine} system index.
+     *
+     * @return The {@link SystemIndexDescriptor} for the {@link Engine} system index.
+     */
+    public static SystemIndexDescriptor getSystemIndexDescriptor() {
+        return SystemIndexDescriptor.builder()
+            .setIndexPattern(ENGINE_INDEX_NAME_PATTERN)
+            .setPrimaryIndex(ENGINE_CONCRETE_INDEX_NAME)
+            .setDescription("Contains auth token data")
+            .setMappings(getIndexMappings())
+            .setSettings(getIndexSettings())
+            .setAliasName(ENGINE_ALIAS_NAME)
+            .setVersionMetaKey("version")
+            .setOrigin(ENGINE_ORIGIN)
+            .setThreadPools(ExecutorNames.DEFAULT_SYSTEM_INDEX_THREAD_POOLS)
+            .build();
+    }
+
+    private static Settings getIndexSettings() {
+        return Settings.builder()
+            .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+            .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
```

Comment:
What's the best practice here, setting replicas to 1 or 0? What are the risks of defaulting to 0?

Reply:
It's an old trick, it is set to 0 to ensure that the index is green for clusters with a single data node. The settings below `index.auto_expand_replicas` will expand the number of replicas to 1 based on the number of data nodes.

Reply:
Oh OK, thanks for the explanation!